mm2src/lp_swap/taker_swap.rs

```diff
@@ -1517,20 +1579,31 @@ impl TakerSwap {
             let secret = self.r().secret.0;
             let maker_coin_swap_contract_address = self.r().data.maker_coin_swap_contract_address.clone();
 
-            let transaction = try_s!(
-                self.maker_coin
-                    .send_taker_spends_maker_payment(
-                        &maker_payment,
-                        self.maker_payment_lock.load(Ordering::Relaxed) as u32,
-                        other_maker_coin_htlc_pub.as_slice(),
-                        &secret,
-                        maker_coin_htlc_keypair.private().secret.as_slice(),
-                        &maker_coin_swap_contract_address,
-                    )
-                    .compat()
-                    .await
+            let fut = self.maker_coin.send_taker_spends_maker_payment(
+                &maker_payment,
+                self.maker_payment_lock.load(Ordering::Relaxed) as u32,
+                other_maker_coin_htlc_pub.as_slice(),
+                &secret,
+                maker_coin_htlc_keypair.private().secret.as_slice(),
+                &maker_coin_swap_contract_address,
             );
 
+            let transaction = match fut.compat().await {
+                Ok(t) => t,
+                Err(err) => {
+                    if let Some(tx) = err.get_tx() {
+                        broadcast_p2p_tx_msg(
```

Comment:
Despite the initial requirement and `send_taker_fee` comment, it's ok to `broadcast_p2p_tx_msg` in `recover_funds`. Please keep it here.